beluga_ros/include/beluga_ros/ndt_ellipsoid.hpp

```diff
@@ -0,0 +1,132 @@
+  const auto eigenvalues = eigenSolver.eigenvalues();
+
+  Eigen::Matrix3d rotationMatrix;
+  Eigen::Vector3d scalevector;
+  rotationMatrix << eigenvectors.col(0), eigenvectors.col(1), eigenvectors.col(2);
+  scalevector = Eigen::Vector3d{eigenvalues.x(), eigenvalues.y(), eigenvalues.z()};
+
+  // Permutation
+  if (std::abs(rotationMatrix.determinant() - 1.0) > Eigen::NumTraits<double>::dummy_precision()) {
+    rotationMatrix << eigenvectors.col(1), eigenvectors.col(0), eigenvectors.col(2);
+    scalevector = Eigen::Vector3d{eigenvalues.y(), eigenvalues.x(), eigenvalues.z()};
+  }
+
+  Eigen::Quaterniond rotation{1.0f, 0.0f, 0.0f, 0.0f};
+  if (std::abs(rotationMatrix.determinant() - 1.0) < Eigen::NumTraits<double>::dummy_precision()) {
+    rotation = Eigen::Quaterniond{rotationMatrix};
+  }
+
+  marker.pose.position.x = mean[0];
+  marker.pose.position.y = mean[1];
+  marker.pose.position.z = mean[2];
+
+  marker.pose.orientation.x = rotation.x();
+  marker.pose.orientation.y = rotation.y();
+  marker.pose.orientation.z = rotation.z();
+  marker.pose.orientation.w = rotation.w();
+
+  marker.scale.x = scalevector.x();
+  marker.scale.y = scalevector.y();
+  marker.scale.z = scalevector.z();
+}
+
+static void use_cell_size(const auto& position, const auto& size, auto& marker){
+  marker.type = beluga_ros::msg::Marker::CUBE;
+  marker.action = beluga_ros::msg::Marker::ADD;
+
+  marker.pose.position.x = position[0];
+  marker.pose.position.y = position[1];
+  marker.pose.position.z = position[2];
+
+  marker.scale.x = size;
+  marker.scale.y = size;
+  marker.scale.z = size;
+}
+
+/// Assign an ellipsoid to each cell of a SparseValueGrid. A cube is used instead if the distribution of the
+/// cell is not suitable for the rotation matrix creation.
+ /** 
+  * \param grid A SparseValueGrid that contains cells representing obstacles.
+  * \return A message with the ellipsoids or cubes.
+  * \tparam MapType Container that maps from Eigen::Vector<int, NDim> to the type of the cell. See [SparseValueGrid]
+  * (https://ekumen-os.github.io/beluga/packages/beluga/docs/_doxygen/generated/reference/html/classbeluga_1_1SparseValueGrid.html).
+  * \tparam NDim Dimension of the grid.
+  */ 
+template <typename MapType, int NDim>
+beluga_ros::msg::MarkerArray assign_obstacle_map(const beluga::SparseValueGrid<MapType, NDim>& grid) {
+  // Get data from the grid
+  auto& map = grid.data();
+
+  // Clean up the message
+  beluga_ros::msg::MarkerArray message{};
+  {
+    beluga_ros::msg::Marker marker;
+    marker.ns = "obstacles";
+    marker.action = beluga_ros::msg::Marker::DELETEALL;
+    message.markers.push_back(marker);
+  }
+
+  // Add the markers
+  int idCount = 0;
+  for (const auto& [key, cell] : map) {
+    beluga_ros::msg::Marker marker;
+    marker.header.frame_id = "map";
+    marker.id = idCount++;
+    marker.ns = "obstacles";
+
+    const auto eigenSolver = Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, 3, 3>>{cell.covariance};
```

Comment:
@Fernando-Sanz consider instantiating the solver once, using `Eigen::SelfAdjointEigenSolver::compute` on each covariance matrix.